Envoy: SO_LINGER redux

jrajahalme opened the pull request:
Add a new `--envoy-http-upstream-linger-timeout` option, and a helm value `envoy.httpUpstreamLingerTimeout` for controlling SO_LINGER config on Envoy HTTP listeners. Helm value defaults to not being set, which makes the agent to default to value `-1`, causing no SO_LINGER config to be set.

Value `0` causes upstream HTTP connections reusing the original source address and port of the source pod to be closed immediately when calling `close()`. This leads to TCP RST frames being sent on the network, but makes original source address and port of the connection immediately available for reuse.

Positive values make Envoy worker threads block during the `close()` system call for the duration it takes to complete TCP FIN/ACK sequence. If the given number of seconds passes, the connection will be reset using TCP RST. Note that blocking the worker thread will have an impact on Envoy throughput.

Cilium CI passed with both `0` and `1` settings, but we still opted to default to `-1` so that there is no change in behavior without explicit opt-in to setting this to either `0` or `1`. Higher values are discouraged due to the worker thread blocking issue.

Please note that any number of network and/or cluster effects could cause the TCP close sequence to take time, including network policy changes that would drop the TCP FIN frames, or the destination otherwise becoming unresponsive.

```release-note
An option was added to control SO_LINGER config on Envoy HTTP upstream connections that bind to the pod's original source address and port. This can be used to mitigate bind errors that could happen if the upstream HTTP connection can not be cleanly closed immediately.
```

jrajahalme committed:
envoy: Set socket linger config on HTTP listeners

Set socket linger config to 0 seconds, to be applied to upstream Envoy
connections for HTTP filter chains that also use the original source
address and port. This should help avoid bind errors on new upstream
connection when an earlier socket with the same source address and port
is still lingering.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
vendor: Update API for BpfMetadata filter update

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras commented on images/cilium/Dockerfile:
```diff
@@ -6,7 +6,7 @@ ARG CILIUM_RUNTIME_IMAGE=quay.io/cilium/cilium-runtime:e65085ae25146e444896e13c5
 # cilium-envoy from github.com/cilium/proxy
 #
-ARG CILIUM_ENVOY_IMAGE=quay.io/cilium/cilium-envoy:v1.33.1-1742784275-e76b29a8d9694b2a6a015b6b9bb82a450bf51451@sha256:062e832f5cbbacf7690e8fe669db6abe12ba45e5b53fbffee37762713d5ab51e
+ARG CILIUM_ENVOY_IMAGE=quay.io/cilium/cilium-envoy:v1.33.1-d7c01ba8c491cc337eccfbb831ad40fb4c6f2c4a@sha256:17d9dc8034d4c3ee342c6d78b799b24bb0d5a343df81be6c0359408e7683b673
```
can you use the one with timestamp tag, so that renovate can continue doing auto upgrade?

```
quay.io/cilium/cilium-envoy:v1.33.1-1742980217-d7c01ba8c491cc337eccfbb831ad40fb4c6f2c4a@sha256:17d9dc8034d4c3ee342c6d78b799b24bb0d5a343df81be6c0359408e7683b673
```

sayboras replied:
The related script is updated in https://github.com/cilium/cilium/pull/38544

jrajahalme replied:
Thanks for fixing the script!

In the meanwhile main had picked up the latest cilium/proxy main, so no need for `cilium-envoy` image update in the PR.

mhofstetter commented on pkg/ciliumenvoyconfig/cec_resource_parser.go:
```diff
@@ -464,8 +464,23 @@ func (r *cecResourceParser) parseResources(cecNamespace string, cecName string,
 				port := uint16(listener.GetAddress().GetSocketAddress().GetPortValue())
+				useZeroLinger := false
+				// Check if this is a HTTP listener
+			Loop:
```
please don't use labels - instead we should introduce a separate method that does the check whether linger should be used or not (with an early return)

sayboras replied:
second the above :+1:

jrajahalme replied:
Replaced the loop label use with a function as suggested :-)

jrajahalme replied:
(function, but not a method).

marseel reviewed (approved):
vendor changes lgtm

sayboras reviewed (approved):
One comment as per above, the rest looks reasonable.

jrajahalme committed:
envoy: Add envoy-http-upstream-linger-timeout option

Make envoy upstream socket linger timeout configurable, defaults to -1
(no change in behavior).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
